### PR TITLE
Help Center: Happychat HE availability

### LINKED ISSUE #63815: Task: Implement Happychat socket connection in the Help Center
### Details

It seems we can’t check the Happychat availability unless we connect to the `WebSocket` server of Happychat (source: pdtkmj-9r-p2 - search for "socket"). The good news is that it turns out Happychat uses JWT for auth, and sends the token in-session, meaning it doesn’t use cookies at all. So, we’ll probably be able to check the status even with 3rd party cookies are blocked. Unless, [we won't be able to get sign the JWT token in that context](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/happychat/utils.js#L8), we'll see.

What we’ll need to do is the following (I already spiked and all seems to work):

- [x] Move the [Happychat connection manager](https://github.com/Automattic/wp-calypso/blob/36fcd301d0fde08f136b6cb14ae4c451a3162d37/client/lib/happychat/connection.js) into a package. The only Calypso dependencies it has are [these action creators](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/happychat/connection.js#L4-L17). I think we can pass them in the constructor of the `Connection` class. And make them optional like so `dispatch( receiveConnect?.() )` because we won't need most if them in the help center.
- [x] Move as much as possible of [this helper](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/happychat/utils.js#L12) to the same package, it's entirely entangled with Calypso state, so copying the logic but not the code is OK.
- [x] Create a hook `useHCAvailable()` to check for [the `accept` value](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/happychat/connection.js#L70). If you agree with passing callback in the constructor of `Connection`, we'll only need to pass `receiveAccept`, as this function will give you if Happychat is available at that moment.
- [x] Once the hook is ready, we can use it to decide whether to show the chat option in the help center.

/cc @klimeryk 

### Checklist

_No response_

### Related

_No response_

### PR DESCRIPTION
**Pull request by escapemanuele**
#### Changes proposed in this Pull Request

- [x] Moved the [Happychat connection manager](https://github.com/Automattic/wp-calypso/blob/36fcd301d0fde08f136b6cb14ae4c451a3162d37/client/lib/happychat/connection.js) into a package.
- [x] Created a hook `useHappychatAuth` for getting user auth
- [x] Created a hook `useHappychatAvailable` that with the auth check to see if there are chat slots available
- [x] Used the last hook to set the Live chat availability

TODO: We have to consider if we need [skills and groups](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/happychat/utils.js#L17-L24), and if so implement them.

#### Testing instructions

* Checkout branch.
* Run `yarn test-client lib/happychat` and confirm that all tests succeed.
* Test that the chat still works

Related to #63815
Fixes #63815

### CODE REVIEW

File: packages/happychat-connection/src/connection.ts [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+		this.receiveLocalizedSupport = receiveLocalizedSupport;
+		this.receiveMessage = receiveMessage;
+		this.receiveMessageOptimistic = receiveMessageOptimistic;
+		this.receiveMessageUpdate = receiveMessageUpdate;
+		this.receiveReconnecting = receiveReconnecting;
+		this.receiveStatus = receiveStatus;
+		this.receiveToken = receiveToken;
+		this.receiveUnauthorized = receiveUnauthorized;
+		this.requestTranscript = requestTranscript;
+	}
```
Comment (alshakero) [inline packages/happychat-connection/src/connection.ts]:
```suggestion
	constructor( props: ConnectionProps = {} ) {
		Object.assign(this, props);
	}
```

File: packages/happychat-connection/src/use-happychat-available.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+	useEffect( () => {
+		if ( ! isLoadingAuth && dataAuth ) {
+			const connection = buildConnection( {
+				receiveAccept: ( receivedAvailability ) => setIsAvailable( receivedAvailability ),
+			} );
+			connection.init( ( value: unknown ) => value, Promise.resolve( dataAuth ) );
+		}
+	}, [ dataAuth, isLoadingAuth ] );
+
+	return { available, isLoading: available === undefined };
```
Comment (alshakero) [inline packages/happychat-connection/src/use-happychat-available.ts]: Nit: might make the return type cleaner.

```suggestion
	return { available: Boolean( available ), isLoading: available === undefined };
```

**Review by alshakero — COMMENTED** [review]
Amazing work, truly.

I'll test everything tomorrow.

**Comment by escapemanuele** [comment]
Thank you @alshakero. Still some typing and import problems to be addressed tho, but should be some easy fix 👍

File: packages/happychat-connection/src/connection-async.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	receiveLocalizedSupport,
+	receiveMessage,
+	receiveMessageOptimistic,
+	receiveMessageUpdate,
+	receiveReconnecting,
+	receiveStatus,
+	receiveToken,
+	receiveUnauthorized,
+	requestTranscript,
+}: ConnectionProps ) {
```
Comment (alshakero) [inline packages/happychat-connection/src/connection-async.ts]:
```suggestion
export default function buildConnection( props: ConnectionProps ) {
```

File: packages/happychat-connection/src/connection-async.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+					receiveLocalizedSupport,
+					receiveMessage,
+					receiveMessageOptimistic,
+					receiveMessageUpdate,
+					receiveReconnecting,
+					receiveStatus,
+					receiveToken,
+					receiveUnauthorized,
+					requestTranscript,
+				} )
```
Comment (alshakero) [inline packages/happychat-connection/src/connection-async.ts]:
```suggestion
				createConnection( props )
```

**Review by alshakero — COMMENTED** [review]
Tiny comments until I give this a thorough review.

File: packages/happychat-connection/src/use-happychat-auth.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import wpcomRequest from 'wpcom-proxy-request';
+import type { HappychatSession, HappychatUser, User } from './types';
+
+type Result = {
+	url: string;
+	user: {
+		jwt: string;
+	} & HappychatUser;
+};
+
```
Comment (alshakero) [inline packages/happychat-connection/src/use-happychat-auth.ts]: This looks exactly like `HappychatAuth` interface.

File: packages/happychat-connection/src/connection.ts [hunk 6]
```diff
@@ -1,33 +1,39 @@
...
+	receiveMessage?: ( message: string ) => void;
+	receiveMessageOptimistic?: ( message: string ) => void;
+	receiveMessageUpdate?: ( message: string ) => void;
+	receiveReconnecting?: () => void;
+	receiveStatus?: ( status: string ) => void;
+	receiveToken?: () => void;
+	receiveUnauthorized?: ( message: string ) => void;
+	requestTranscript?: () => void;
+	dispatch: any;
+	openSocket: any;
```
Comment (alshakero) [inline packages/happychat-connection/src/connection.ts]: I'm not sure about this suggestion, but here goes: 

```suggestion
	dispatch: (arg: Action) => void;
	openSocket: Socket | undefined;
```

File: packages/happychat-connection/src/use-happychat-available.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+import useHappychatAuth from './use-happychat-auth';
+
+export function useHappychatAvailable() {
+	const [ available, setIsAvailable ] = useState< boolean | undefined >( undefined );
+	const { data: dataAuth, isLoading: isLoadingAuth } = useHappychatAuth();
+
+	useEffect( () => {
+		if ( ! isLoadingAuth && dataAuth ) {
+			const connection = buildConnection( {
+				receiveAccept: ( receivedAvailability ) => setIsAvailable( receivedAvailability ),
```
Comment (alshakero) [inline packages/happychat-connection/src/use-happychat-available.ts:12]: I think we should add `receiveUnauthorized` here and set `available` to false, if it's called.

**Review by alshakero — APPROVED** [review]
Great great work! Calypso works and code looks great. Left some optional comments.

**Pull request merged**